File: opendev/core/utils/sound.py
"""Sound utility for task notifications."""

import os
import platform
import subprocess
import logging
import time

logger = logging.getLogger(__name__)

_last_played: float = 0.0
_COOLDOWN_SECONDS: float = 30.0

def play_finish_sound():
    """Play a sound to indicate task completion.
    
    Tries various system-specific methods to play a sound.
    Fails silently if no method works.
    """
    global _last_played
    now = time.monotonic()
    if now - _last_played < _COOLDOWN_SECONDS:
        return
    _last_played = now

    system = platform.system()
    try:
        if system == "Darwin":  # macOS
            # Glass is a nice built-in sound
            subprocess.Popen(["afplay", "/System/Library/Sounds/Glass.aiff"], 
                            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        elif system == "Linux":
            # Try common Linux sound players
            players = ["paplay", "aplay", "play", "cvlc"]
            # Common locations for sounds on Linux
            sounds = [
                "/usr/share/sounds/freedesktop/stereo/complete.oga",
                "/usr/share/sounds/gnome/default/alerts/glass.ogg",
                "/usr/share/sounds/alsa/Front_Center.wav"
            ]
            
            played = False
            for player in players:
                # Check if player exists
                try:
                    if subprocess.run(["which", player], capture_output=True, shell=False).returncode == 0:
                        for sound in sounds:
                            if os.path.exists(sound):
                                if player == "cvlc":
                                    subprocess.Popen([player, "--play-and-exit", sound], 
                                                    stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
                                else:
                                    subprocess.Popen([player, sound], 
                                                    stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
                                played = True
                                break
                except Exception:
                    continue
                if played:
                    break
            
            if not played:
                # Fallback to system bell
                print("\a", end="", flush=True)
                
        elif system == "Windows":
            # Use PowerShell to play a system sound
            # We use double backslashes for the path and escape them for the shell if needed
            sound_path = "C:\\Windows\\Media\\notify.wav"
            ps_cmd = f"(New-Object Media.SoundPlayer '{sound_path}').PlaySync()"
            subprocess.Popen(["powershell", "-Command", ps_cmd], 
                            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        else:
            # Fallback for other systems
            print("\a", end="", flush=True)
    except Exception as e:
        # Ignore errors if sound cannot be played
        logger.debug(f"Failed to play finish sound: {e}")
        pass

File: crates/opendev-runtime/src/sound.rs
//! Sound utility for task completion notifications.
//!
//! Plays a system sound when a task completes. Platform-aware:
//! - macOS: `afplay` with Glass sound
//! - Linux: tries `paplay`, `aplay`, `play`, `cvlc` with common sound files
//! - Other: terminal bell (`\a`)
//!
//! Includes a cooldown to prevent rapid repeated sounds.

use std::sync::atomic::{AtomicU64, Ordering};
use std::time::Instant;

use tracing::debug;

/// Minimum seconds between consecutive sounds.
const COOLDOWN_SECONDS: u64 = 30;

/// Monotonic timestamp of the last played sound (epoch millis approximation).
static LAST_PLAYED_MS: AtomicU64 = AtomicU64::new(0);

/// Lazy-initialized start time for monotonic clock.
static START: std::sync::OnceLock<Instant> = std::sync::OnceLock::new();

fn now_ms() -> u64 {
    let start = START.get_or_init(Instant::now);
    start.elapsed().as_millis() as u64
}

/// Play a sound to indicate task completion.
///
/// Fails silently if no sound player is available.
/// Respects a 30-second cooldown between plays.
pub fn play_finish_sound() {
    let now = now_ms();
    let last = LAST_PLAYED_MS.load(Ordering::Relaxed);
    if now.saturating_sub(last) < COOLDOWN_SECONDS * 1000 {
        return;
    }
    LAST_PLAYED_MS.store(now, Ordering::Relaxed);

    if let Err(e) = play_platform_sound() {
        debug!("Failed to play finish sound: {e}");
    }
}

fn play_platform_sound() -> Result<(), String> {
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("afplay")
            .arg("/System/Library/Sounds/Glass.aiff")
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .spawn()
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    #[cfg(target_os = "linux")]
    {
        let players = ["paplay", "aplay", "play", "cvlc"];
        let sounds = [
            "/usr/share/sounds/freedesktop/stereo/complete.oga",
            "/usr/share/sounds/gnome/default/alerts/glass.ogg",
            "/usr/share/sounds/alsa/Front_Center.wav",
        ];

        for player in &players {
            let which = std::process::Command::new("which").arg(player).output();
            if let Ok(output) = which
                && output.status.success()
            {
                for sound in &sounds {
                    if std::path::Path::new(sound).exists() {
                        let mut cmd = std::process::Command::new(player);
                        if *player == "cvlc" {
                            cmd.arg("--play-and-exit");
                        }
                        cmd.arg(sound)
                            .stdout(std::process::Stdio::null())
                            .stderr(std::process::Stdio::null())
                            .spawn()
                            .map_err(|e| e.to_string())?;
                        return Ok(());
                    }
                }
            }
        }

        // Fallback: terminal bell
        print!("\x07");
        Ok(())
    }

    #[cfg(target_os = "windows")]
    {
        let sound_path = r"C:\Windows\Media\notify.wav";
        let ps_cmd = format!("(New-Object Media.SoundPlayer '{sound_path}').PlaySync()");
        std::process::Command::new("powershell")
            .args(["-Command", &ps_cmd])
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .spawn()
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    #[cfg(not(any(target_os = "macos", target_os = "linux", target_os = "windows")))]
    {
        print!("\x07");
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cooldown_logic() {
        // Test that the cooldown mechanism works
        let now = now_ms();
        assert!(now >= 0);

        // If LAST_PLAYED is set to now, subsequent calls within 30s should be blocked
        LAST_PLAYED_MS.store(now, Ordering::Relaxed);

        let new_now = now_ms();
        let last = LAST_PLAYED_MS.load(Ordering::Relaxed);
        // Within cooldown window
        assert!(new_now.saturating_sub(last) < COOLDOWN_SECONDS * 1000);
    }

    #[test]
    fn test_now_ms() {
        let t1 = now_ms();
        let t2 = now_ms();
        assert!(t2 >= t1);
    }
}
